core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -616,84 +619,108 @@ class KafkaApis(val requestChannel: RequestChannel,
-              new TopicMetadata(topic, Seq.empty[PartitionMetadata], Errors.INVALID_TOPIC_EXCEPTION.code)
-          }
+        if (topic == GroupCoordinator.GroupMetadataTopicName) {
+          createGroupMetadataTopic()
+        } else if (config.autoCreateTopicsEnable) {
+          createTopic(topic, config.numPartitions, config.defaultReplicationFactor)
         } else {
-          new TopicMetadata(topic, Seq.empty[PartitionMetadata], Errors.UNKNOWN_TOPIC_OR_PARTITION.code)
+          new MetadataResponse.TopicMetadata(Errors.UNKNOWN_TOPIC_OR_PARTITION, topic, java.util.Collections.emptyList())
         }
       }
-      topicResponses.appendAll(responsesForNonExistentTopics)
+      topicResponses ++ responsesForNonExistentTopics
     }
-    topicResponses
   }
 
   /**
    * Handle a topic metadata request
    */
   def handleTopicMetadataRequest(request: RequestChannel.Request) {
-    val metadataRequest = request.requestObj.asInstanceOf[TopicMetadataRequest]
-
-    //if topics is empty -> fetch all topics metadata but filter out the topic response that are not authorized
-    val topics = if (metadataRequest.topics.isEmpty) {
-      val topicResponses = metadataCache.getTopicMetadata(metadataRequest.topics.toSet, request.securityProtocol)
-      topicResponses.map(_.topic).filter(topic => authorize(request.session, Describe, new Resource(Topic, topic))).toSet
+    val metadataRequest = request.body.asInstanceOf[MetadataRequest]
+
+    val topics = metadataRequest.topics.asScala.toSet
+    var (authorizedTopics, unauthorizedTopics) = if (metadataRequest.topics.isEmpty) {
+      //if topics is empty -> fetch all topics metadata but filter out the topic response that are not authorized
+      val authorized = metadataCache.getAllTopics()
+        .filter(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
+      (authorized, mutable.Set[String]())
     } else {
-      metadataRequest.topics.toSet
+      topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
     }
 
-    //when topics is empty this will be a duplicate authorization check but given this should just be a cache lookup, it should not matter.
-    var (authorizedTopics, unauthorizedTopics) = topics.partition(topic => authorize(request.session, Describe, new Resource(Topic, topic)))
-
-    if (!authorizedTopics.isEmpty) {
-      val topicResponses = metadataCache.getTopicMetadata(authorizedTopics, request.securityProtocol)
-      if (config.autoCreateTopicsEnable && topicResponses.size != authorizedTopics.size) {
-        val nonExistentTopics: Set[String] = topics -- topicResponses.map(_.topic).toSet
+    if (authorizedTopics.nonEmpty) {
+      val nonExistingTopics = metadataCache.getNonExistingTopics(authorizedTopics)
+      if (config.autoCreateTopicsEnable && nonExistingTopics.nonEmpty) {
         authorizer.foreach {
           az => if (!az.authorize(request.session, Create, Resource.ClusterResource)) {
```

Comment:
This is existing code, but the formatting is a bit weird, `az` should be in the previous line.